core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Example.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.concurrent.Future
+
+import akka.Done
+import com.softwaremill.react.kafka_api_producer.Producer
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+
+trait Example {
+  implicit val ec = scala.concurrent.ExecutionContext.Implicits.global
+
+  def consumerProvider: () => KafkaConsumer[Array[Byte], String] = ???
+
+  def producerProvider: () => KafkaProducer[Array[Byte], String] = ???
+
+  class DB {
+    def save(offset: Long, data: String): Future[Done] = ???
+  }
+}
+
+// Consume messages and store a representation, including offset, in DB
+object Example3 extends Example {
+  val db = new DB
+
+  Consumer.source(consumerProvider)
```

Comment:
It seems like GH formatted my previous comment wrong, so there is one question missing from me:

> Not emitting onComplete may dead lock your stream (with splitAfter for example). 

I don't see any deadlock here. Can you build an example that deadlocks? The onComplete will be emitted, just after last message has been committed (at least in the scenario I described).

> I was talking about case when your process can be shutdown in reasonable time. Sure I should be ready for hanging, but also I want to have correct predictable stream shutdown if everything goes ok to minimize data lost.

Then I propose the following external `stop()` semantics:
 - `stop()` is called externally
 - Source stops emitting new elements
 - Waits until all messages are committed, or configurable timeout fires
 - on timeout do an onFailure otherwise an onComplete on the stream
 - release the Kafka Consumer